### PR TITLE
libfdt: overlay: Refactor overlay_fixup_phandle and update test case

### PR DESCRIPTION
**Pull request by Drupping**
This commit makes two main changes:
1. Refactored overlay_fixup_phandle to optimize efficiency by moving the phandle lookup logic based on label outside the overlay_fixup_one_phandle call. This avoids redundant phandle lookups when a single label is associated with multiple modifications.

2. Updated the test case for overlay_bad_fixup. Changed the target DTS from overlay_base_no_symbols.test.dtb to overlay_base_manual_symbols.test.dtb. This ensures that the test case doesn't exit prematurely due to the absence of label-linked phandle in the __symbols__ node. The update guarantees that the test case appropriately checks the validity of the fixup string linked to the label, as intended.

### CODE REVIEW

File: libfdt/fdt_overlay.c [hunk 1]
```diff
@@ -404,6 +392,18 @@ static int overlay_fixup_phandle(void *fdt, void *fdto, int symbols_off,
...
+		return prop_len;
+	
+	symbol_off = fdt_path_offset(fdt, symbol_path);
+	if (symbol_off < 0)
+		return symbol_off;
+	
+	phandle = fdt_get_phandle(fdt, symbol_off);
+	if (!phandle)
+		return -FDT_ERR_NOTFOUND;
+
```
Comment (dgibson) [inline libfdt/fdt_overlay.c:406]: This is a nice refactor, thanks!.

File: libfdt/fdt_overlay.c [hunk 2]
```diff
@@ -444,8 +444,8 @@ static int overlay_fixup_phandle(void *fdt, void *fdto, int symbols_off,
 			return -FDT_ERR_BADOVERLAY;
 
 		ret = overlay_fixup_one_phandle(fdt, fdto, symbols_off,
-						path, path_len, name, name_len,
-						poffset, label);
+										path, path_len, name, name_len,
+										poffset, phandle);
```
Comment (dgibson) [inline libfdt/fdt_overlay.c:448]: These lines are oddly over-indented in the revised version, can you fix this up please.

File: tests/run_tests.sh [hunk 3]
```diff
@@ -269,7 +269,7 @@ libfdt_overlay_tests () {
     for test in $BAD_FIXUP_TREES; do
 	tree="overlay_bad_fixup_$test"
 	run_dtc_test -I dts -O dtb -o $tree.test.dtb "$SRCDIR/$tree.dts"
-	run_test overlay_bad_fixup overlay_base_no_symbols.test.dtb $tree.test.dtb
+	run_test overlay_bad_fixup overlay_base_manual_symbols.test.dtb $tree.test.dtb
```
Comment (dgibson) [inline tests/run_tests.sh:272]: This change is unrelated, logically or code-wise to the other one.  Can you please make them separate patches / commits (they can still go in the same PR on github).

**Review by dgibson — CHANGES_REQUESTED** [review]
Sorry it's taken me so long to look at these.  The idea of the change looks good, but there are a couple of small nits I'd like improved before merge.

File: tests/run_tests.sh (hunk 3 above)
Comment (Drupping) [reply tests/run_tests.sh:272]: The test case update is necessary due to the refactoring, as it's essential for the tests to pass with the modified code logic. Specifically, the change prevents premature test exits due to the updated phandle handling. For the detailed rationale, please refer to point 2 in my original commit message. Nonetheless, I've split the commits for clarity as you suggested.

Comment (dgibson) [reply tests/run_tests.sh:272]: Sorry, I missed that.  I had the impression that the first patch didn't make any behavioural change, just optimised the existing behaviour.  Since it clearly does, I need to consider more carefully if this is the right idea.

I'm having trouble understanding point 2 from your commit message.  Can you clarify for me why the behaviour is different with the overlay change?

Comment (Drupping) [reply tests/run_tests.sh:272]: To clarify point two, the key difference between `overlay_base_manual_symbols.test.dtb` and `overlay_base_no_symbols.test.dtb` is that the former includes a `__symbols__` node which is crucial for our updated logic.

I'd like to draw attention to this code segment:

```c
symbol_off = fdt_path_offset(fdt, symbol_path);
if (symbol_off < 0)
	return symbol_off;
```
This block's role is to locate the `__symbols__` node in the base fdt. Before refactoring, the code checked the format of fixup string(like "/fragment@0:target:ab") in the fdto before looking up the `__symbols__` in the base fdt. After refactoring, we first check for the `__symbols__` node in the base fdt.

Therefore, post-refactor, the `__symbols__` node must exist in the base fdt to ensure the code evaluates fixup strings correctly and doesn't exit early due to missing `__symbols__`.

I hope this clarification is helpful. Let me know if further details are needed.

Comment (Drupping) [reply tests/run_tests.sh:272]: Apologies for the mix-up. The code block I previously posted was incorrect. Here’s the correct one.
```c
/* And base DTs without symbols */
symbols_off = fdt_path_offset(fdt, "/__symbols__");
if ((symbols_off < 0 && (symbols_off != -FDT_ERR_NOTFOUND)))
	return symbols_off;
```
Also, my earlier explanation was not entirely accurate. Simply put, we need the base fdt to contain a `__symbols__` node so that the test case will process the logic for checking the fixup string format. Without this, the test would prematurely exit with an FDT_ERR_NOTFOUND error at ```symbol_path = fdt_getprop(fdt, symbols_off, label, &prop_len);```

Comment (dgibson) [reply tests/run_tests.sh:272]: Ah, I see.  Previously the generated trees here had two defects: first they were missing the symbols, and second the actual defect we introduced deliberately to test.  With your other change we're now failing on that first defect instead of the second one we're trying to test.

**Comment by dgibson** [comment]
Thanks for the clarifications.  I've now merged this - I reversed the order of the two patches so that there isn't a temporary test failure when bisecting.
